### ISSUE TITLE
[Feature Request] parallel downloads feature using a batch file

### ISSUE THREAD

**Issue by Rush2088** [ytdl-org/youtube-dl#3746]
Is there a way to enable parallel download session fed in by a --batch-file.
Some sites have very slow band width per video, therefore instead of waiting to finish previous file, it is much efficient to run parallel downloads.

**Comment by megapctr** [comment]
:+1: It's doable with external downloaders, but that's an annoying hurdle

**Comment by schmod** [comment]
Is there any reason why `xargs` is not be sufficient for this?

**Comment by yan12125** [comment]
> Is there any reason why ```xargs``` is not be sufficient for this?

Some cases: custom output templates, postprocessing...

**Comment by Roman2K** [comment]
I would rather have the complexity of parallel downloads kept out of youtube-dl so that its code base remains more maintainable.

Parallelism can be already be delegated to `xargs -P x -n y` where `x` is the number of parallel downloads and `y` is the number of URLs/IDs to pass to youtube-dl, most likely `1`. I find this a very elegant solution, composing various tools, each one doing a single job well. Used it successfully many times, saving me hours or days.

@yan12125 
> Some cases: custom output templates, postprocessing...

Maybe I'm missing something but why can't this be done with `xargs youtube-dl` passing the same arguments (for templates, postprocessing, ...) as you would `youtube-dl`?

**Comment by yan12125** [comment]
From the viewpoint of youtube-dl internals, a series of urls is not different than a playlist with several videos. I was considering the latter case when I wrote the previous comment. If you want to achieve downloading videos of a playlist in parallel along with postprocessing them, you'll need to run the parallel download step - either via internal codes or external tools - from youtube-dl, instead of run youtube-dl in parallel, and then the postprocessing step can continue inside youtube-dl.

Or moving postprocessing codes to standalone commands. That would be interesting.

**Comment by siddht4** [comment]
the biggest bottleneck is that there is no exit code which defines wheter a download occured or not.It has to be done manually.

Though this can be achived by setting some sort of flag (true if downloaded,false if any error), It may create file overwrite.

So another method is copying the batch file internally as list, splitting them evenly , then running parallel download for list_1 , list_2.

**Comment by siddht4** [comment]
After bit of research and all,I found out that https://github.com/MrS0m30n3/youtube-dl-gui   already does supports it.

https://pypi.python.org/pypi/twodict  may help.

https://github.com/MrS0m30n3/youtube-dl-gui/blob/master/youtube_dl_gui/downloadmanager.py#L106-L122  manages the current state.  

https://github.com/MrS0m30n3/youtube-dl-gui/blob/master/youtube_dl_gui/downloadmanager.py#L239-L385 manages the queue which can be modified to work in parallel.


So it seems the process already exist so a little tinkering will add to the main project.

**Comment by eladkarako** [comment]
<h2><sub><sup>A Workable Solution For A </sup></sub>parallel download <sub><sup>(<strong>beta</strong>),</sup></sub></h2>

<h2>testers needed!</h2>

<ol>
<li><a href="https://github.com/eladkarako/youtube-dl-p/archive/master.zip">download</a> and extract.</li>
<li>Add <code>youtube-dl-p-master</code> folder to your system-<code>PATH</code>. <sup><em>optional</em></sup></li>
<li>Copy few video-links (from YouTube) to a text file.</li>
<li>run <code>ytdl your-text-file.txt</code>.</li>
</ol>

Unlike `--external-downloader aria2c`, it will parallel download the entire list.

<a href="https://github.com/eladkarako/youtube-dl-p/issues/new">issues</a>.

**Comment by eladkarako** [comment]
@siddht4 get the download data using youtube-dl (<code>--skip-download --dump-json</code>), write <sup><a href="https://github.com/eladkarako/youtube-dl-p/tree/master/ytdl2aria">#1</a></sup> an <a href="https://aria2.github.io/manual/en/html/aria2c.html#cmdoption-i">aria2 compatible-list</a>, and rely on aria2 for <strong>both the queue-management</strong> and download <sup><a href="https://github.com/eladkarako/youtube-dl-p/">#2</a></sup>, to get both multipart-download and multi-file download.

<sub>For example, aria2 with <code>--split=3 --min-split-size=1M --max-concurrent-downloads=5 --max-connection-per-server=16</code> will give you an active download-queue of five files, each has three parallel segment-download at a time. works well on youtube.</sub>

**Comment by alex-hofmann** [comment]
Just wanted to thank @Roman2K and provide the full one-liner, since many youtube-dl users may not be familiar with xargs or the shell. 
> Parallelism can be already be delegated to `xargs -P x -n y` where `x` is the number of parallel downloads and `y` is the number of URLs/IDs to pass to youtube-dl, most likely `1`. I find this a very elegant solution, composing various tools, each one doing a single job well. Used it successfully many times, saving me hours or days.

`cat files.txt | xargs -n 1 -P 4 youtube-dl`
Here, `files.txt` is my batch file; -n 1 indicates one url for each youtube-dl call; -P 4 indicates 4 parallel youtube-dl calls

**Comment by aryehbeitz** [comment]
@alex-hofmann can this be used to download in parallel when downloading a channel or playlist?

**Comment by PiotrDabrowskey** [comment]
@aryehbeitz No, I guess. It only allows you to download multiple links in parallel, but you need to specify those links in the file. Playlists and channels are just single link which are passed to youtube-dl as you normally do, so it does not solve the problem of this issue.

If there's a way to gather all the links from one playlist and place them into the file, this would do the trick.

**Comment by alex-hofmann** [comment]
@aryehbeitz @PiotrDabrowskey Yeah I think you're right, this is only useful when you can feed the known links. For example, if someone wanted to download the Jeopardy! episodes that a reddit user graciously uploads daily they could use something like
`curl -A 'random' https://www.reddit.com/user/jibjabjrjr/.rss 2>/dev/null | grep -Po 'https:\/\/drive.*?(?=\;|\&)' | xargs -n 1 -P 4 youtube-dl`

My guess is that something similar is available for pulling the urls from a youtube playlist with a little curl+grep

**Comment by Roman2K** [comment]
@aryehbeitz You can get the individual URLs from a playlist with:
```sh
$ youtube-dl -j --flat-playlist 'https://www.youtube.com/watch?v=CcNo07Xp8aQ&list=PLA671B7E7BFB780B7' | jq -r '"https://youtu.be/"+ .url'
```
That can be fed to directly to `xargs` for parallel downloads as in @alex-hofmann's example.

**Comment by Roman2K** [comment]
If anyone's interested, I made a cronjob for backing up my YouTube playlists to Google Drive (using rclone). Idempotent/incremental. Kind of rough scripting but I've been running it with several 500+ long playlists over the last few days and it works well.

* Basic usage: `ruby dl.rb PLAYLIST_URL`
  * downloads to `./out`
  * keeps metadata/cache in `./meta`
* Export to Drive/other: see `dl_to_rclone`
  * usage: see `favorites/dl.sh`
* Cronjob: see `crontab` and `cronjob`
  * depends on [alerterr](https://github.com/Roman2K/alerterr)
* Concurrency: see `Downloader::NTHREADS` in dl.rb

Clone and customize to your needs.

**Comment by zer0def** [comment]
Trashman's oneliner (requires `jq` alongside typical `coreutils` binaries):
```
youtube-dl -j -o "%(autonumber)s-%(title)s.%(ext)s" -u <username> -p <password> <playlist_url> | xargs -I'{}' -d'\n' -n1 -P`nproc` -- /bin/sh -c 'TMP=`mktemp`; echo '\''{}'\'' >${TMP}; youtube-dl -f best -o "`jq -r ._filename ${TMP}`" --load-info ${TMP} &>/dev/null; rm ${TMP}'
```

**Comment by epitron** [comment]
@zer0def Nice!

Small issue: the number of parallel downloaders is set to the number of CPU cores (-P\`nproc\`). That parameter should be tuned to match the network capacity, not the CPU capacity.

Anyone who wants to try that should start it around -P5, and increase it until it saturates their connection.

**Comment by zer0def** [comment]
@epitron As long as your extractions don't fall back to ffmpeg, it's safe to ramp up the process number, but when they do, you might end up unnecessarily extending your downloads due to threads fighting for CPU time, so I've picked `nproc` simply because it's fairly idiot-proof, while remaining faster than a sequential run on a large playlist.

As per my previous comment: adjust to your heart's content, since everyone's case might be different.

**Comment by epitron** [comment]
@zer0def Hyperthreading can also artificially increase your CPU count. What if you have 32 cores, and a regular network connection?

I thought YTDL only used ffmpeg to copy streams into different containers (unless you pass `--recode-video` or a specific format that needs to be transcoded). Are there any services that will by-default transcode video, and which support playlists?

**Comment by zer0def** [comment]
As mentioned above, it's a **safe** number (meaning good enough™) to which there are always going to be exceptions. Even poorly done HT (looking at you, AMD Bulldozer) works well enough™. If you end up having **more threads spawned than your network connection can handle**, I'd argue that by virtue of being orders of magnitude slower, saturating out your connection is preferred, even if it means your threads will be stalling for data.
With that said, this is splitting hairs and if you're capable of knowing better, you'll be making min-maxing adjustments based on your specific conditions anyway. This nitpicking is exactly the case.

As for the latter paragraph, as an example, this is what you frequently (at least half of the time) get fetching content behind JWPlayer on LinuxAcademy:
```
WARNING: hlsnative has detected features it does not support, extraction will be delegated to ffmpeg
```

**Comment by zer0def** [comment]
Then you have threads stalling for data, which is hardly a concern compared to saturating out low-end and/or low-core count CPUs with a pipe huge enough to overwhelm it.

As an example, this is what you frequently (at least half of the time) get fetching content behind JWPlayer on LinuxAcademy:
```
WARNING: hlsnative has detected features it does not support, extraction will be delegated to ffmpeg
```

**Comment by zer0def** [comment]
Then you have threads stalling for data, which is hardly a concern compared to saturating out low-end and/or low-core count CPUs with a pipe huge enough to overwhelm it.

**Comment by epitron** [comment]
@zer0def What if YouTube bans or throttles you for spamming their servers with connections?

**Comment by zer0def** [comment]
Do you have an actual case to provide as basis or are we just talking hypotheses for the sake of doing so?

Besides, what's wrong with getting throttled if the service provider (in this particular case, Youtube) deems so to provide adequate service to their consumers? It's within the provider's right to enforce their ToS as strictly or loosely as they choose. If they cared enough, they'd simply make using ytdl miserable.

Do mind that "adjust to your heart's content" (which you insist on ignoring, btw) also asserts "don't be dumb", so if one is running a Threadripper and is still mindlessly copying lengthy oneliners they might not understand from a page labeled "parallel downloads with youtube-dl™" on the 'net, then they're effectively asking to get banned/throttled for being potentially harmful through ignorance. And that's fine.

So far you're trying to make a golden mean argument to an inherently scattershot problem, to which the solution would go outside the "trashman's" point of having as little dependencies as necessary. If you're still looking, `parallel`'s got you covered. Or your favorite programming language.

**Comment by zer0def** [comment]
Do you have an actual case to provide as basis or are we just talking hypotheses? Because if you're looking for a golden mean, then yes, you could just cobble up a proper solution, write a wrapper script or use `parallel --limit`. All of which go beyond a minimal dependencies "trashman's" case.

**Comment by siddht4** [comment]
As this feature has not been implemented I would suggest you to embed youtube-dl and perform multiprocessing or multithreading. Whichever suits you best. For that to try split your batch file aka url to half split . Run 2 parallel process using kworker and check wheter its feasible for you or not.          Youtube does blocks multiple request so I would suggest not to overload using youtube-dl bot.

**Comment by siddht4** [comment]
I have built a youtube-dl  embed code which can help to parallel process the user input file

[https://github.com/siddht4/youtube-dl-embed/blob/master/python/multi_processing.py](url)

Here I have manually selected the file in line 43 
`path='./list/list.txt'`

which can be changed accordingly.

The file list will be used as the input .

If the list is greater than 1 , then two parallel process will start where the list will first be splitted into half and then the process will start the half half list independently  [https://github.com/siddht4/youtube-dl-embed/blob/master/python/multi_processing.py#L110-L115](url)

`        _a1,_a2=split_list(lines)
        # send the half half list to processor so that two parallel process will  start and simulatneously download two url 
        processor(_a1,_a2)`


For single url two parallel process will start speeding the download    [https://github.com/siddht4/youtube-dl-embed/blob/master/python/multi_processing.py#L103-L107](url)


Would suggest to include a parameter     `-mp 2` in youtube-dl itself so the multiprocessing is included as a parameter itself. 

There is also way to use multiple cores instead of multiple threads   [https://github.com/siddht4/youtube_dl_custom_old/blob/master_1/multithreading/main.py#L38-L40](url)

Source for multi core  [https://github.com/siddht4/youtube_dl_custom_old/blob/master_1/multithreading/main.py](url)

**Comment by dirkf** [comment]
The original question was answered (tldr: yes, but) with no specific changes required. Linked issues discuss parallelisation within the program.
